Postgres installation and configuration

jeanleonov opened the pull request:
Should go after https://github.com/AppScale/appscale-thirdparties/pull/1.

jeanleonov committed:
Postgres installation and configuration

sjones4 commented on postgres/install.sh:
```diff
@@ -0,0 +1,22 @@
+
+log "Installing Postgres"
+attempt=1
+while ! (yes | apt-get install postgresql)
```
```suggestion
while ! apt-get --assume-yes install postgresql
```

sjones4 commented on postgres/configure-and-start-postgres.sh:
```diff
@@ -0,0 +1,142 @@
+# Let's get the command line arguments.
+while [ $# -gt 0 ]; do
+    if [ "${1}" = "--host" ]; then
+        shift; if [ -z "${1}" ]; then usage; fi
```
This test will fail when `${1}` does not exist (due to shift), i.e. use `${1:-}`

sjones4 commented on postgres/configure-and-start-postgres.sh:
```diff
@@ -0,0 +1,142 @@
+PG_MAJOR_VER=$(psql --version | awk '{ print $3 }' | awk -F '.' '{ print $1 }')
+PG_VERSION=$(psql --version | awk '{ print $3 }' | awk -F '.' '{ print $1 "." $2 }')
+
+if [ "${PG_MAJOR_VER}" = "10" ]; then
```
Should this be >= 10?

sjones4 commented on postgres/configure-and-start-postgres.sh:
```diff
@@ -0,0 +1,142 @@
+# Allow host connections to the specified DB
+if grep -q -E "^host[ \t]+${DBNAME}[ \t]+${USERNAME}[ \t]+" "${PG_HBA}"
+then
+    sed -i "s|^host[ \t]+${DBNAME}[ \t]+${USERNAME}[ \t]+.*|host ${DBNAME} ${USERNAME} ${HOST}/0 md5|" "${PG_HBA}"
```
Should this be `0.0.0.0/0` rather than `${HOST}/0`?

jeanleonov replied:
You're right. I copied this mistake from TaskQueue e2e test. Probably initial intention was to limit access to the server from outside private network, but then we'd need to use mask like 24.
I'll change it to `0.0.0.0/0` for now as mask of 24 bits won't be working for all networks.

sjones4 commented on postgres/configure-and-start-postgres.sh:
```diff
@@ -0,0 +1,142 @@
+    exit 0
+else
+    rm ~/.pgpass
+fi
```
You can avoid password files and using sudo by making use of `pg_ident.conf` to map `root` to `postgres`.

in `pg_ident.conf`: 
`rootmap         root                    postgres`

update `pg_hba.conf`: 
`local   all             postgres                                peer map=rootmap`

then access as postgres using `psql -U postgres` or `createdb -U postgres`.

jeanleonov replied:
Thanks, I didn't know about `rootmap` feature.
The reason why `.pgpass` file is used is to test if specified database was previously created and specified user can login with provided password (so to test that user will be able to connect using corresponding DSN string).
I see that `rootmap` would also help to get rid of `sudo -u`, but I'd prefer to leave it as it is for now.
Let me know if you think it's critical.

jeanleonov committed:
Addressing PR comments

 - Fix args parsing issue.
 - Use major version in configs path starting from v10.
 - Use 0.0.0.0/0 to match connection source.

jeanleonov committed:
Remove data-dir customization

jeanleonov committed:
Run apt-get update before installing postgres

cdonati commented on postgres/configure-and-start-postgres.sh:
```diff
@@ -0,0 +1,122 @@
+### Actual installation procedure ###
+#####################################
+
+${SCRIPT_DIR}/install.sh
```
Does the start script need to run install? If so, could it check if it's already installed? I could imagine a situation where we'd want to be able to start postgres without an internet connection, and `apt-get update` would fail in that case.

cdonati commented on postgres/configure-and-start-postgres.sh:
```diff
@@ -0,0 +1,122 @@
+if psql --dbname ${DBNAME} --username ${USERNAME} --host ${HOST} \
+        --command 'SELECT current_timestamp;'
+then
+    rm ~/.pgpass
```
You may want this in an exit trap just in case more commands are added after creation of the file (since -e is set).

jeanleonov committed:
Check if fdb and postgres are installed

 - better handling of public address customization.

sjones4 commented:
Daily build no. 6926
